mono/metadata/handle.h

```diff
@@ -547,6 +552,64 @@ mono_context_get_handle (void);
 void
 mono_context_set_handle (MonoAppContextHandle new_context);
 
+/*
+mono_unwrapped_string_t is a common representation
+for both a MonoString* and a MonoStringHandle.
+
+This would be much simpler if we are using C++ -- constructor, destructor, overload, etc.
+
+Usage patterns:
+1.     MonoStringHandle s;
+       mono_unwrapped_string_t us = { 0 };
+       ...
+       us = mono_unwrap_string_handle (s);
+       us.length;
+       us.chars;
+       mono_unwrapped_string_cleanup (&us); // ok without mono_unwrap_string_handle
+
+2.     MonoStringHandle s;
+       mono_unwrapped_string_t us = mono_unwrap_string_handle (s);
+       us.length;
+       us.chars;
+       mono_unwrapped_string_cleanup (&us);
+
+3.     MonoString* s;
+       mono_unwrapped_string_t us = mono_unwrap_string (s);
+       us.length;
+       us.chars;
+       mono_unwrapped_string_cleanup (&us); // optional
+
+4.     MonoString* s;
+       mono_unwrapped_string_t us = { 0 ];
+       ...
+       us = mono_unwrap_string (s);
+       us.length;
+       us.chars;
+       mono_unwrapped_string_cleanup (&us); // optional, ok without mono_unwrap_string_handle
+*/
+typedef struct mono_unwrapped_string_t {
```

Comment:
I find having some unambiguous small typographic convention very useful but ok. Otherwise is subtlety around functions being verbs and types being nouns.

Reply:
One thing I don't like about the _Foo convention is the names leak into any mangled C++ names, and cannot be changed (i.e. removed if converted to C++) w/o breaking dynamic link compatibility. We might not have all extern C but the types can leak and be used by public.